File: mediuim/majority_element_ii.rb
# Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times. The algorithm should run in linear time and in O(1) space.
# 
# Hint:
# 
# How many majority elements could it possibly have?

# @param {Integer[]} nums
# @return {Integer[]}
def majority_element(nums)
  len = nums.length
  n1, n2, c1, c2 = nil, nil, 0, 0
    
  nums.each do |num|
    if num == n1
      c1 += 1
    elsif num == n2
      c2 += 1
    elsif c1 == 0
      n1,c1 = num,1
    elsif c2 == 0
      n2,c2 = num,1
    else
      c1 -= 1
      c2 -= 1
    end
  end
  
  result = []
  result << n1 if nums.count(n1) > len/3
  result << n2 if nums.count(n2) > len/3
  result
end

File: mediuim/majority_element_ii.js
var majorityElement = function(nums) {
    var buckets = [];
    var size = nums.length;
    if(size == 0) return [];
    if(size <= 3) return nums;

    for(var i=0; i < size; i++){
        var num = nums[i];
        if(!buckets[0] || buckets[0].count === 0){
            buckets[0] = { number: num, count: 1 };
        }
        else{
            if(buckets[0].number === num) buckets[0].count++;
            else {
                buckets[0].count--;
                if(buckets[1] && buckets[1].count > 0){
                    if(buckets[1].number === num) buckets[1].count++;
                    else buckets[1].count--;
                }
                else buckets[1] = { number:num, count:1 };
            }
        }
    }

    return buckets
             .map(function(b){ return b.number; })
	     .filter(function(n){
               var numberCount = 0;
	       for(var j=0; j<size; j++){
                 if(n === nums[j]) numberCount++;
               }
	       return numberCount > size/3.0;
	     });
}
